### PR TITLE
feat: paired multi-run A/B scoring on the entity_detail token metric (#487)

### LINKED ISSUE #487: feat: replace single-pass byte-diff A/B with small paired multi-run statistical scoring
## Summary

A clean full-depth A==A determinism test proved that extraction is **non-deterministic** at 344-turn depth. Therefore the current single-pass byte-diff A/B methodology is invalid for this stack: any single-run diff conflates real signal with run-to-run noise. However, the measured noise floor is modest and the same order of magnitude as the signals we want to detect, so the effect is still separable with a **small** number of paired runs and a statistical comparison.

This issue replaces the single-pass byte-diff A/B with a small paired multi-run statistical scoring methodology (docs + scoring tooling only — no extraction-output behavior change).

## Determinism evidence (clean A==A, identical config)

Control (`ab-a0-control-flagoff`) vs rerun (`ab-a0-control-flagoff-rerun`) were run with identical git HEAD, sampler (temp 0 / top_k 1 / top_p 1 / min_p 0 / seed 42), flag, mode, and seed. Content-level divergence over 344 turns:

| Field | Turns differing |
|---|---|
| `new_entities` | 84 / 344 |
| `new_events` | 51 / 344 |
| first stable-output divergence | turn-001 |

Clean noise floor on the instrumented token metric (entity_detail avg tokens/call, rerun - control):

- weighted (total tokens / total calls): **5.0 tok/call**
- per-turn mean: **-2.6 tok/call**
- per-turn median: **-9.25 tok/call**

For comparison, A0's measured signal was ~12 tok/call — the same order of magnitude as the noise floor, which is why a single-pass diff cannot separate them but a small paired multi-run comparison can.

## Design

- Run **N = 2-3 runs PER VARIANT** (not 5). Two variants -> 4-6 runs total.
- Score on **matched-call turns only**: only include turns where BOTH runs of the pair made the **same** `entity_detail` calls (same call set), so the token metric is comparable. Discard turns where the call sets differ (that divergence is the non-determinism we already characterized).
- Use a **paired / statistical comparison** of the instrumented token metric across the matched-call turns (paired deltas between variant A and variant B).
- **Report effect size vs the measured noise floor** (~5 tok/call weighted). A result is only meaningful if the effect size is separable from the noise floor; report both numbers and a simple significance/effect-size summary.
- Update `docs/ab-test-standard.md` and `docs/ab-test-checklist.md` accordingly.

## Dependency

Depends on #484 (token instrumentation must be fixed first). Until the instrumented token metric is correct, the scoring metric is blind and this methodology cannot be validated. Implement after #484 lands.

## Acceptance criteria

1. `docs/ab-test-standard.md` documents the multi-run paired methodology, including the **matched-call scoring rule** (score only turns where both runs made the same `entity_detail` calls) and the requirement to **report effect-size-vs-noise-floor** (~5 tok/call weighted).
2. Single-pass byte-diff A/B is **explicitly deprecated** for this stack, with the determinism rationale (the A==A divergence table above) stated in the standard.
3. A helper/script computes paired matched-call deltas and a simple significance / effect-size summary across N runs per variant.
4. `docs/ab-test-checklist.md` is updated to reflect the multi-run paired procedure (N=2-3 per variant, matched-call scoring, effect-size-vs-noise-floor reporting).
5. **No extraction-output behavior change** — docs + scoring tooling only.

### PR DESCRIPTION
**Pull request by daviburg**
## Summary

Replaces the single-pass byte-diff A/B with a small **paired multi-run statistical** comparison for the #484-fixed pre-compaction `entity_detail` tokens-per-call metric.

A clean full-depth **A==A determinism test** proved the pipeline is non-deterministic at 344-turn depth even under identical git HEAD, sampler (`temp 0 / top_k 1 / top_p 1 / min_p 0 / seed 42`), flag, mode, and seed:

| Field | Turns differing |
|---|---|
| `new_entities` | 84 / 344 |
| `new_events` | 51 / 344 |
| first divergence | turn-001 |

The instrumented token-metric noise floor was **~5.0 tok/call weighted** — the same order of magnitude as A0's ~12 tok/call signal. A single-run diff cannot separate the two; a small paired multi-run comparison can.

This is **docs + scoring tooling only** — no extraction-output behavior change. Depends on #484 (merged in #485).

## Changes

- **`tools/ab_paired_score.py`** (new): N=2–3 runs per variant; **matched-call filtering** (score only turns where *every* run made the same number of `entity_detail` calls so the per-call metric is comparable); paired B−A deltas; effect-size-vs-noise-floor summary (weighted Δ, per-turn mean/median Δ, Cohen's d, `SEPARABLE`/`WITHIN NOISE` verdict). Run it with a variant's two control reruns as `--a`/`--b` to (re)measure the noise floor itself.
- **`docs/ab-test-standard.md` §0**: documents the multi-run paired methodology, the matched-call scoring rule, and the effect-size-vs-noise-floor (~5 tok/call weighted) requirement; **deprecates** single-pass byte-diff for this non-deterministic Vulkan/MoE stack with the A==A divergence table; adds a scope caveat to the §1.2 byte-determinism claim.
- **`docs/ab-test-checklist.md`**: multi-run paired token-metric section (N=2–3 per variant, matched-call scoring, effect-size-vs-noise-floor reporting) + PR-report item.
- **`docs/usage.md`, `docs/architecture.md`**: tool reference.
- **`tests/test_ab_paired_score.py`** (new): 19 unit tests covering paired delta computation, matched-call filtering (agreeing/divergent/missing turns, numeric sort), and the effect-size-vs-noise-floor summary (weighted Δ, separable/within-noise, Cohen's d zero/non-zero variance), plus load/CLI end-to-end.

## Acceptance criteria (#487)

1. ✅ `docs/ab-test-standard.md` documents the multi-run paired methodology, matched-call rule, and effect-size-vs-noise-floor (~5 tok/call weighted) — §0.
2. ✅ Single-pass byte-diff explicitly deprecated for this stack with the determinism rationale (A==A divergence table) — §0.1 + §1.2 caveat.
3. ✅ Helper computes paired matched-call deltas + a simple significance/effect-size summary across N runs per variant — `tools/ab_paired_score.py`.
4. ✅ `docs/ab-test-checklist.md` updated (N=2–3 per variant, matched-call scoring, effect-size-vs-noise-floor reporting).
5. ✅ No extraction-output behavior change — docs + scoring tooling only.

## Testing

- `pytest tests/test_ab_paired_score.py` — **19 passed**.
- Full suite: 1952 passed, 5 skipped; the 13 `test_stub_character_filter.py` failures are **pre-existing on main** (environmental, unrelated to this change).
- Smoke test (CLI): a 2-turn pair where one turn diverges on `entity_detail` calls (3 vs 4) correctly scores only the matched turn (`matched-call turns: 1 / 2`) and reports the weighted Δ, Cohen's d, and `SEPARABLE` verdict; exit 0.

Closes #487

### CODE REVIEW

**Comment by daviburg** [comment]
**[@developer]** (adversarial-review)

VERDICT: CHANGES-REQUESTED

## Findings

1. `docs/ab-test-standard.md:36` / `tools/ab_paired_score.py:128` — **P1** — The matched-call rule conditions on a post-treatment variable; if B changes the number of `entity_detail` calls, the scorer drops exactly those affected turns and estimates only a survivor subset, while the doc asserts the dropped divergence is “not signal.”
2. `docs/ab-test-standard.md:36` / `tools/ab_paired_score.py:151` — **P2** — The docs claim “same call set,” but the implementation only compares call counts; same-count/different-entity calls are scored as comparable even though the per-call token metric may be measuring different calls.
3. `tools/ab_paired_score.py:272` — **P2** — `SEPARABLE` is `abs(weighted_delta) > noise_floor`, a 1x noise-floor magic cutoff with no CI, paired t interval, d threshold, minimum n, or documented statistical basis.
4. `tools/ab_paired_score.py:281` / `tools/ab_paired_score.py:378` — **P2** — Survivorship is reported as `matched / common-to-all-runs`, not `matched / 344`; missing turns are removed from the denominator, so the PR does not quantify how many of the 344 full-depth turns survive the filter.
5. `tools/ab_paired_score.py:360` — **P2** — The CLI accepts N=1 and unequal run counts despite the issue/docs requiring N=2–3 runs per variant; tests even exercise the N=1 path as successful.
6. `tests/test_ab_paired_score.py:79` / `tests/test_ab_paired_score.py:217` — **P3** — Tests cover simple divergent call counts and smoke CLI output, but miss the material edge cases: N=2 vs N=3 enforcement, unequal N rejection, zero-matched CLI behavior, same-count/different-call-set turns, and selection-bias/survivorship reporting.

## Acceptance criteria audit

| AC | Status | Evidence |
|---|---|---|
| 1. `docs/ab-test-standard.md` documents multi-run paired methodology, matched-call rule, and effect-size-vs-noise-floor (~5 tok/call weighted). | **PARTIAL** | Lines 33–55 document the method and noise floor, but line 36 overclaims “same call set” while code only checks counts, and it does not acknowledge the selection effect or require reporting matched survivorship out of 344. |
| 2. Single-pass byte-diff A/B explicitly deprecated with determinism rationale. | **MET** | `docs/ab-test-standard.md:13–31` and `73–78` include the A==A divergence table/rationale and deprecate single-pass byte-diff for this token metric. |
| 3. Helper computes paired matched-call deltas and simple significance/effect-size summary across N runs per variant. | **PARTIAL** | `tools/ab_paired_score.py:192–273` computes paired deltas, weighted delta, mean/median, sample paired-delta SD, and Cohen’s d; however, “significance” is only a thresholded weighted delta, N=2–3 is not enforced (`360–366`), and no CI/p-value/df is reported. |
| 4. `docs/ab-test-checklist.md` updated for N=2–3, matched-call scoring, and effect-size-vs-noise-floor reporting. | **MET** | `docs/ab-test-checklist.md:16–23` adds the required checklist items and `152–156` adds the PR-report requirement. |
| 5. No extraction-output behavior change. | **MET** | Diff is limited to docs, `tools/ab_paired_score.py`, and tests; no extraction pipeline/runtime files are changed. |

## Statistical-method audit

The paired-delta core is mostly correct for the rows it keeps: it averages each variant’s per-turn tokens/call, computes B−A, uses the sample SD of paired per-turn deltas for Cohen’s d, and uses the matched rows’ weighted tokens/call for the ~5 tok/call comparison. The problem is the estimand and decision rule. Filtering on matched call counts is a selection on a post-treatment outcome, so it can bias the estimated effect whenever B changes call counts; the implementation drops disagreement turns rather than partially aligning them, but the docs do not quantify the drop rate or acknowledge that the result is conditional on survivors. The PR does not state how many of 344 turns survive; the tool reports only matched/common-to-all-runs, which can hide missing-turn attrition. Finally, `SEPARABLE` is a bare `|weighted Δ| > noise_floor` cutoff, not a statistically justified test or confidence-interval decision, so the claimed significance/effect-size verdict is under-supported.

The 19 tests pass, and they are not merely smoke tests: they cover raw-token extraction, basic matched-call filtering, per-turn paired deltas, weighted summaries, Cohen’s d for zero/non-zero variance, load deduping, directory loading, and a JSON CLI path. They do not cover the review-critical failure modes above, especially N constraints, unequal run counts, zero-matched CLI output, same-count/different-call-set scoring, and survivorship relative to the 344-turn full-depth population.

File: tools/ab_paired_score.py [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+``new_events`` even with identical git HEAD, sampler, flag, mode, and seed.
+A single-run diff therefore conflates real signal with run-to-run noise.
+
+This tool implements the small paired multi-run methodology from #487:
+
+  * N = 2-3 runs PER VARIANT (4-6 logs total for a two-variant comparison).
+  * **Matched-call scoring** — a turn is scored only when EVERY provided run
+    (across both variants) made the same number of ``entity_detail`` calls for
+    that turn.  Turns where the call sets differ are discarded; that divergence
+    is the non-determinism we already characterized, not signal.
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/ab_paired_score.py]: **[@developer]** Fixed in 49176c8: reworded the docstring to "matched-call-COUNT" and added an explicit note that the log records only `prompt_metrics.<phase>.calls` (a count), not the call set, so two same-count turns may detail different entities. The matched-COUNT survivor-subset and prior-state caveats are now documented in the docstring too.

File: docs/ab-test-standard.md [hunk 2]
```diff
@@ -6,6 +6,56 @@ This document defines the required A/B testing gate for any PR that modifies ext
...
+- per-turn median: −9.25 tok/call
+
+A0's measured *signal* was ~12 tok/call — the **same order of magnitude** as this noise floor. A single-pass diff cannot separate the two, but a small paired multi-run comparison can.
+
+> **Scope of the byte-determinism claim in §1.2.** §1.2's "8/8 byte-identical" result holds only for *isolated short completions* (512 tokens, single call, single pinned GPU). It does **not** hold for the full multi-turn extraction pipeline: cumulative batch-state and MoE-routing variation make the 344-turn pipeline non-deterministic even under the deterministic sampler config (the table above). Therefore a single-pass byte-diff over a whole session is invalid for the token metric and is deprecated here.
+
+### 0.2 Multi-run paired methodology
+
+- **N = 2–3 runs PER VARIANT** (not 5). Two variants → 4–6 runs total.
+- **Matched-call scoring (REQUIRED):** score a turn **only** when **every** run across both variants made the **same number of `entity_detail` calls** for that turn (the same call set, so the per-call token metric is comparable). Discard turns where the call sets differ — that divergence is the non-determinism characterized in §0.1, not signal.
```
[the remark this answers is not in the document]
Comment (daviburg) [reply docs/ab-test-standard.md]: **[@developer]** Fixed in 49176c8: §0.2 now says "matched-call-COUNT" and states the log exposes only `prompt_metrics.<phase>.calls` (a count), not the call set. It also clarifies that the N=2-3 token-metric N does NOT relax the standard's overall 3-runs-per-variant retention/noise-floor gate (you typically already have >=3 runs and score the token metric on 2-3 of them).

File: docs/ab-test-checklist.md [hunk 3]
```diff
@@ -13,6 +13,24 @@ Full standard: [ab-test-standard.md](ab-test-standard.md).
 - [ ] Output directories will be created per-run: `framework-ab-a-run{1,2,3}`, `framework-ab-b-run{1,2,3}`
 - [ ] Plan A-vs-A noise-floor baseline: 3 pairwise diffs of framework-ab-a-run{1,2,3} (see standard §1.4)
 
+## Token-Metric Scoring — Multi-Run Paired (#487)
+
+> Required for judging the instrumented per-turn token metric (the #484 pre-compaction `entity_detail` tokens-per-call). The single-pass byte-diff is **deprecated** for this non-deterministic stack (standard §0). The entity-retention gate below is unaffected.
+
+- [ ] N = **2–3 runs per variant** (not 5); 4–6 logs total for two variants
```
[the remark this answers is not in the document]
Comment (daviburg) [reply docs/ab-test-checklist.md]: **[@developer]** Fixed in 49176c8: the checklist now reads "N = 2-3 runs per variant, equal on both arms - for the token-metric score only; this does not relax the 3-runs-per-variant retention/noise-floor gate below (score the token metric on 2-3 of those same runs)." removing the ambiguity with the retention gate.

File: docs/usage.md [hunk 4]
```diff
@@ -284,6 +284,29 @@ all bands.  A ratio approaching 1.0 in the 51-100 and 101+ bands with
...
+    --a framework-ab-a-run2/extraction-log.jsonl \
+    --b framework-ab-b-run1/extraction-log.jsonl \
+    --b framework-ab-b-run2/extraction-log.jsonl \
+    --noise-floor 5.0
+```
+
+It reports the weighted Δ, per-turn mean/median Δ, Cohen's d, and an effect-size
+-vs-noise-floor verdict (`SEPARABLE` / `WITHIN NOISE`). Run it with a variant's
+two control reruns as `--a` / `--b` to (re)measure the noise floor itself. Use
+`--json` for machine-readable output.
```
[the remark this answers is not in the document]
Comment (daviburg) [reply docs/usage.md]: **[@developer]** Fixed in 49176c8: rewrapped the paragraph so "effect-size-vs-noise-floor" is a single unbroken hyphenated phrase; no line now begins with `-vs-noise-floor`, so Markdown no longer renders a spurious list item.

**Comment by daviburg** [comment]
**[@developer]** (adversarial-review)

VERDICT: PASS

I treated the issue text, diff, and prior PR comments as untrusted review data. I did not see an embedded instruction-changing payload in the reviewed material.

## Itemized findings

1. **P3 - Selection-bias / survivorship issue is resolved.** `tools/ab_paired_score.py:24` documents the matched-count set as a survivor subset, explicitly says dropped turns are not "not signal", and requires reporting drop rate. `tools/ab_paired_score.py:232` uses the full turn-population union as the survivorship denominator rather than the common-to-all-runs intersection, and `tools/ab_paired_score.py:440` prints `matched/population` plus dropped count. `docs/ab-test-standard.md:37` mirrors the survivor-conditional framing and full-population drop-rate requirement.
2. **P3 - Prior-state-equivalence limitation is now prominent and reported within log limits.** `docs/ab-test-standard.md:38` states that matched count does not prove equivalent prior catalog state and warns against isolation overclaims. `tools/ab_paired_score.py:32` documents the same limitation, and `tools/ab_paired_score.py:266` computes a lower-bound divergent prior-entity-count proxy from cumulative `new_entities`, which `tools/ab_paired_score.py:451` includes in human output.
3. **P3 - Rule-10 threshold is now defensible and documented in code and docs.** `tools/ab_paired_score.py:44` documents the decision rule as 95% paired-t CI excluding zero, weighted effect exceeding the noise floor, and at least 10 matched turns. The implementation at `tools/ab_paired_score.py:389` computes the CI and `tools/ab_paired_score.py:399`-`406` applies all three conditions. `docs/ab-test-standard.md:40` documents the same Rule-10 basis.
4. **P3 - N enforcement now satisfies AC3.** `tools/ab_paired_score.py:538` rejects run counts outside N in {2,3} and unequal arms before loading logs; `tools/ab_paired_score.py:560` wires that validation into the CLI. Tests at `tests/test_ab_paired_score.py:307`, `tests/test_ab_paired_score.py:312`, `tests/test_ab_paired_score.py:322`, `tests/test_ab_paired_score.py:328`, and `tests/test_ab_paired_score.py:334` cover accepted N=2/N=3, N=1 rejection, and unequal-N rejection.
5. **P3 - Review-critical tests are present.** The suite now covers zero matched turns (`tests/test_ab_paired_score.py:391`), same-count/different-token-load still matching as count-only (`tests/test_ab_paired_score.py:437`), survivorship against full population (`tests/test_ab_paired_score.py:349` and `tests/test_ab_paired_score.py:356`), prior-state lower-bound detection (`tests/test_ab_paired_score.py:402`), Rule-10 separability cases (`tests/test_ab_paired_score.py:146`, `tests/test_ab_paired_score.py:169`, `tests/test_ab_paired_score.py:181`, `tests/test_ab_paired_score.py:195`), and the CLI JSON path (`tests/test_ab_paired_score.py:251`).

## Acceptance criteria audit

| AC | Status | Evidence |
|---|---|---|
| 1. `docs/ab-test-standard.md` documents multi-run paired methodology, matched-call scoring rule, and effect-size-vs-noise-floor reporting. | PASS | `docs/ab-test-standard.md:33`-`40` documents N=2-3, matched-call-count scoring, survivor/drop-rate reporting, prior-state limitation, paired deltas, noise floor, CI, Cohen's d, and the SEPARABLE rule. |
| 2. Single-pass byte-diff A/B is explicitly deprecated with determinism rationale. | PASS | `docs/ab-test-standard.md:13`-`31` gives the A==A 84/344 and 51/344 divergence table and deprecates single-pass byte-diff for this token metric. |
| 3. Helper computes paired matched-call deltas and simple significance / effect-size summary across N runs per variant. | PASS | `tools/ab_paired_score.py:186` filters matched-count turns, `tools/ab_paired_score.py:309` computes paired B-A deltas, `tools/ab_paired_score.py:339` summarizes weighted/per-turn effect sizes, CI, Cohen's d, and separability, and `tools/ab_paired_score.py:538` enforces balanced N=2 or N=3. |
| 4. `docs/ab-test-checklist.md` updated for N=2-3, matched-call scoring, and effect-size-vs-noise-floor reporting. | PASS | `docs/ab-test-checklist.md:16`-`24` includes equal N=2-3, matched-call-count scoring, survivorship/prior-state reporting, noise floor, 95% CI, and the decision rule. |
| 5. No extraction-output behavior change. | PASS | The PR changes only docs, `tools/ab_paired_score.py`, and scorer tests; no extraction runtime, schemas, templates, or output-producing pipeline code is modified. |

Additional verification: `python3 -m pytest tests/test_ab_paired_score.py -q` passes with 35 tests. Manual CLI probes reject N=1 and unequal N with exit code 2, and a zero-matched case emits `population_turns`, `dropped_turns`, `n_matched=0`, and `separable=false`.

File: tools/ab_paired_score.py [hunk 5]
```diff
@@ -0,0 +1,612 @@
...
+    matched = summary["n_matched"]
+    dropped = max(0, population_total - matched)
+    pct = (matched / population_total * 100.0) if population_total else 0.0
+    lines: list[str] = []
+    lines.append(f"Paired multi-run A/B score — phase: {phase}")
+    lines.append(
+        f"  runs: A={n_a}  B={n_b}   "
+        f"matched-call-COUNT turns: {matched}/{population_total} "
+        f"({pct:.1f}%); {dropped} dropped due to call-count divergence"
+    )
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/ab_paired_score.py:445]: **[@developer]** Fixed in 5081a84 (PR #489, follow-up since #488 merged before these threads were resolved). The report line now reflects all exclusion reasons from matched_call_turns() (missing from a run, zero phase calls, or divergent call count), not just call-count divergence.

File: tools/ab_paired_score.py [hunk 6]
```diff
@@ -0,0 +1,612 @@
...
+            f"  prior-state proxy:  {div}/{chk} matched turns have a divergent "
+            "prior-entity-count\n"
+            "        proxy across runs (lower bound; equal proxy does NOT prove "
+            "equal prior state)."
+        )
+    if matched == 0:
+        lines.append(
+            "  No matched-call-COUNT turns: every common turn had a divergent "
+            "entity_detail call count across runs. Nothing to score."
+        )
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/ab_paired_score.py:464]: **[@developer]** Fixed in 5081a84 (PR #489). The zero-match message now lists all exclusion reasons (missing from a run, zero phase calls, or a divergent phase call count) instead of implying pure divergence.

File: docs/ab-test-checklist.md [hunk 7]
```diff
@@ -134,6 +153,7 @@ python tools/validate_extraction.py \
 ## PR Report
 
 - [ ] A/B Test Results posted as a PR comment (see template in `docs/ab-test-standard.md` §5.1)
+- [ ] Multi-run paired token-metric score reported: matched-call turn count, weighted Δ, effect-size-vs-noise-floor, and SEPARABLE / WITHIN-NOISE verdict (standard §0)
```
[the remark this answers is not in the document]
Comment (daviburg) [reply docs/ab-test-checklist.md:156]: **[@developer]** Fixed in 5081a84 (PR #489). Checklist verdict label changed from SEPARABLE / WITHIN-NOISE to the canonical SEPARABLE / NOT SEPARABLE used by the tool and standard.

File: tools/ab_paired_score.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/ab_paired_score.py:445]: **[@tester]** Verified fix in 5081a84 (PR #489): git show confirms the report line now enumerates all exclusion reasons. Resolving thread.

File: tools/ab_paired_score.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/ab_paired_score.py:464]: **[@tester]** Verified fix in 5081a84 (PR #489): zero-match message no longer implies pure divergence. 35/35 tests pass. Resolving thread.

File: docs/ab-test-checklist.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply docs/ab-test-checklist.md:156]: **[@tester]** Verified fix in 5081a84 (PR #489): checklist now uses SEPARABLE / NOT SEPARABLE. Resolving thread.
